File: divide-and-conquer/convert-sorted-array-to-binary-search-tree/convert-sorted-array-to-binary-search-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func sortedArrayToBST(nums []int) *TreeNode {
    return build(nums, 0, len(nums)-1)
}

func build(a []int, lo, hi int) *TreeNode {
    if lo > hi {
        return nil
    }
    mid := lo + (hi-lo)/2
    node := &TreeNode{Val: a[mid]}
    node.Left = build(a, lo, mid-1)
    node.Right = build(a, mid+1, hi)
    return node
}

File: divide-and-conquer/convert-sorted-array-to-binary-search-tree/convert-sorted-array-to-binary-search-tree.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution
{
    public TreeNode SortedArrayToBST(int[] nums)
    {
        return Build(nums, 0, nums.Length - 1);
    }
    
    private TreeNode Build(int[] a, int lo, int hi)
    {
        if (lo > hi) return null;

        int mid = lo + ((hi - lo) >> 1);
        var node = new TreeNode(a[mid]);

        node.left  = Build(a, lo, mid - 1);
        node.right = Build(a, mid + 1, hi);

        return node;
    }
}